xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/PVRClient.h

```diff
@@ -0,0 +1,4763 @@
+  PVR_CONNECTION_STATE_UNKNOWN = 0,
+
+  /// @brief backend server is not reachable (e.g. server not existing or
+  /// network down)
+  PVR_CONNECTION_STATE_SERVER_UNREACHABLE = 1,
+
+  /// @brief backend server is reachable, but there is not the expected type of
+  /// server running (e.g. HTSP required, but FTP running at given server:port)
+  PVR_CONNECTION_STATE_SERVER_MISMATCH = 2,
+
+  /// @brief backend server is reachable, but server version does not match
+  /// client requirements
+  PVR_CONNECTION_STATE_VERSION_MISMATCH = 3,
+
+  /// @brief backend server is reachable, but denies client access (e.g. due
+  /// to wrong credentials)
+  PVR_CONNECTION_STATE_ACCESS_DENIED = 4,
+
+  /// @brief connection to backend server is established
+  PVR_CONNECTION_STATE_CONNECTED = 5,
+
+  /// @brief no connection to backend server (e.g. due to network errors or
+  /// client initiated disconnect)
+  PVR_CONNECTION_STATE_DISCONNECTED = 6,
+
+  /// @brief connecting to backend
+  PVR_CONNECTION_STATE_CONNECTING = 7
+} PVR_CONNECTION_STATE;
+//------------------------------------------------------------------------------
+
+/*!
+  * @brief "C" PVR add-on capabilities.
+  *
+  * Structure used to interface in "C" between Kodi and Addon.
+  *
+  * See @ref AddonPVRCapabilities for description of values.
+  */
+typedef struct PVR_ADDON_CAPABILITIES
+{
+  bool bSupportsEPG;
+  bool bSupportsEPGEdl;
+  bool bSupportsTV;
+  bool bSupportsRadio;
+  bool bSupportsRecordings;
+  bool bSupportsRecordingsUndelete;
+  bool bSupportsTimers;
+  bool bSupportsChannelGroups;
+  bool bSupportsChannelScan;
+  bool bSupportsChannelSettings;
+  bool bHandlesInputStream;
+  bool bHandlesDemuxing;
+  bool bSupportsRecordingPlayCount;
+  bool bSupportsLastPlayedPosition;
+  bool bSupportsRecordingEdl;
+  bool bSupportsRecordingsRename;
+  bool bSupportsRecordingsLifetimeChange;
+  bool bSupportsDescrambleInfo;
+  bool bSupportsAsyncEPGTransfer;
+
+  /*!
+    * @brief **required**\n Count of possible values for PVR_RECORDING.iLifetime.
+    * 0 means lifetime is not supported for recordings or no own value
+    * definition wanted, but to use Kodi defaults of 1..365.
+    */
+  unsigned int iRecordingsLifetimesSize;
+
+  /*!
+    * @brief **optional**\n Array containing the possible values for
+    * PVR_RECORDING.iLifetime. Must be filled if iLifetimesSize > 0
+    */
+  PVR_ATTRIBUTE_INT_VALUE recordingsLifetimeValues[PVR_ADDON_ATTRIBUTE_VALUES_ARRAY_SIZE];
+} ATTRIBUTE_PACKED PVR_ADDON_CAPABILITIES;
+
+//==============================================================================
+/// \ingroup cpp_kodi_addon_pvr_Defs
+/// @brief **PVR add-on capabilities**
+///
+/// If a capability is set to **true**, then the corresponding methods from
+/// @ref kodi::addon::CInstancePVRClient need to be implemented.
+///
+/// Used on @ref kodi::addon::CInstancePVRClient::GetAddonCapabilities().
+///
+//@{
+struct AddonPVRCapabilities
```

Comment:
If this is c++, i would make this a `class`. What's the benefit in having everything `public`? Here and everywhere in this file.

Reply:
Was more to use it easy, I can change to class, the constructors to private and the addon instance class as friend.

Reply:
It's used as a struct - I assume you want to set any member from outside.
So what's the benefit of making those private and have Setters/Getters for each an everyone of them?

Reply:
Im a big fan of data encapsulation, because it makes it possible to change the implementation behind the interface to the data without having to adapt every code line that addresses  the data. Worth the price of having to write (simple inline) setters/getters once.

Reply:
@Alwinus pretty please, no friends!

Reply:
Possible I should just shut up as I don't get the whole context.
We are in an `extern "C"` block here thus classes would keep C++ linkage while structs wouldn't - at least that's what I think.

Reply:
And don't get me wrong, I'm all for data encapsulation - but for entities where the data members _are_ the interface they should be public, which is the case here. Value aggregates are not meant to be abstractions.

Reply:
Yes, it's basically about linkage (name mangling and that stuff).

Reply:
I just talked to @AlwinEsch and I think we can/should remove the `extern C` here as it doesn't really make sense, if you use C++ language elements in it and don't need C interop at the first place (which I think is the case for this part of the file).

My opinion for struct vs. class in this particular case still stands, but as you're the PVR maintainer, so I think it's your call anyway.

Reply:
Have it updated, some structures, e.g. PVR_ADDON_CAPABILITIES is no longer needed as C ++, so I added new function with `GetRecordingsLifetimeValues` to omit the list in it.